apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/thread/ThreadProvider.java

```diff
@@ -31,12 +32,58 @@
     }
 
     public Thread getThreadMetrics() {
+        int newStateThreadCount = 0;
+        int runnableStateThreadCount = 0;
+        int blockedStateThreadCount = 0;
+        int waitingStateThreadCount = 0;
+        int timedWaitingStateThreadCount = 0;
+        int terminatedStateThreadCount = 0;
+
+        ThreadInfo[] threadInfos = threadMXBean.getThreadInfo(threadMXBean.getAllThreadIds(), 0);
+        if (threadInfos != null) {
+            for (ThreadInfo threadInfo : threadInfos) {
+                if (threadInfo != null) {
+                    switch (threadInfo.getThreadState()) {
+                        case NEW:
+                            newStateThreadCount++;
+                            break;
+                        case RUNNABLE:
+                            runnableStateThreadCount++;
+                            break;
+                        case BLOCKED:
+                            blockedStateThreadCount++;
+                            break;
+                        case WAITING:
+                            waitingStateThreadCount++;
+                            break;
+                        case TIMED_WAITING:
+                            timedWaitingStateThreadCount++;
+                            break;
+                        case TERMINATED:
+                            terminatedStateThreadCount++;
+                            break;
+                        default:
+                            break;
+                    }
+                } else {
+                    terminatedStateThreadCount++;
+                }
```

Comment:
The code as a whole looks good. However, there may be some inaccuracies in this place. I found that `getAllThreadIds()` and `getThreadInfo` method both return information about the active thread. However, when the thread is in the NEW and TERMINATED state, the `isAlive()` method will return false, that is, they are inactive. Therefore, in this way, we cannot correctly count the number of threads in the NEW and TERMINATED state.

Reply:
So we need to consider whether we want to remove them.

Reply:
> So we need to consider whether we want to remove them.

`New State` is zero always, but `Terminated State` is non-zero sometimes.

So, I should remove them?

Reply:
If there is no threadInfo, I think we should not run statistics for it.

Reply:
If you remove the counting logic from the else code block, you may find that the `Terminated State` is zero always.

Reply:
> If you remove the counting logic from the else code block, you may find that the `Terminated State` is zero always.

Of course.

So, I remove `Terminated State` and `New State` later.